src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/AttributePresenceFilterNode.cs

```diff
@@ -0,0 +1,295 @@
+            {
+                CustomAttribute customAttribute = reader.GetCustomAttribute(handle);
+                EntityHandle customAttributeConstructorHandle = customAttribute.Constructor;
+                MethodDesc customAttributeConstructorMethodDesc = this._module.GetMethod(customAttributeConstructorHandle);
+                DefType customAttributeDefType = customAttributeConstructorMethodDesc.OwningType as DefType;
+                Debug.Assert(customAttributeDefType != null);
+                string customAttributeTypeNamespace = customAttributeDefType.Namespace;
+                string customAttributeTypeName = customAttributeDefType.Name;
+                // System.Runtime.CompilerServices.NullableAttribute is NEVER added to the table (There are *many* of these, and they provide no useful value to the runtime)
+                if (customAttributeTypeNamespace == "System.Runtime.CompilerServices" && customAttributeTypeName == "NullableAttribute")
+                {
+                    continue;
+                }
+                bool addToTable = false;
+                if (customAttributeTypeNamespace.StartsWith("System.Runtime."))
+                {
+                    addToTable = true;
+                }
+                else if (customAttributeTypeNamespace == "Windows.Foundation.Metadata")
+                {
+                    // Windows.Foundation.Metadata attributes are a similar construct to compilerservices attributes. Add them to the table
+                    addToTable = true;
+                }
+                else if (customAttributeTypeNamespace == "System")
+                {
+                    // Some historical well known attributes were placed in the System namespace. Special case them
+                    if (customAttributeTypeName == "ParamArrayAttribute")
+                    {
+                        addToTable = true;
+                    }
+                    else if (customAttributeTypeName == "ThreadStaticAttribute")
+                    {
+                        addToTable = true;
+                    }
+                }
+                else if (customAttributeTypeNamespace == "System.Reflection")
+                {
+                    // Historical attribute in the System.Reflection namespace
+                    if (customAttributeTypeName == "DefaultMemberAttribute")
+                    {
+                        addToTable = true;
+                    }
+                }
+
+                if (!addToTable)
+                    continue;
+
+                customAttributeEntries.Add(new CustomAttributeEntry
+                {
+                    TypeNamespace = customAttributeTypeNamespace,
+                    TypeName = customAttributeTypeName,
+                    Parent = reader.GetToken(customAttribute.Parent)
+                }); ;
+            }
+            return customAttributeEntries;
+        }
+
+        // Algorithm "xor128" from p. 5 of Marsaglia, "Xorshift RNGs"
```

Comment:
@davidwrighton Do we have/need an entry for this in THIRD-PARTY-NOTICES.TXT at the root of the repo?

Reply:
I thought I had dealt with when I checked this in initially, but this is a bit of a funny case. Its a re-implementation of part of an algorithm from a research paper.

Reply:
I believe this is simply the third C code described on [this](https://en.wikipedia.org/wiki/Xorshift) Wikipedia page.